drivers/staging/media/rpivid/rpivid_h265.c

```diff
@@ -0,0 +1,2075 @@
+	//	   "Mv=%llx, Col=%llx, Stride=%x, Buf=%llx->%llx\n",
+	//	   de->rpi_mvbase, de->rpi_colbase, de->ctx->colmv_stride,
+	//	   de->ctx->colmvbuf.addr, de->ctx->colmvbuf.addr +
+	//	   de->ctx->colmvbuf.size);
+
+	rpivid_hw_irq_active2_irq(dev, &de->irq_ent, cb_phase2, de);
+
+	apb_write(dev, RPI_NUMROWS, de->PicHeightInCtbsY);
+}
+
+static void cb_phase1(struct rpivid_dev *const dev, void *v)
+{
+	struct rpivid_dec_env *const de = v;
+	struct rpivid_ctx *const ctx = de->ctx;
+	int status;
+
+	status = check_status(dev);
+	if (status != 0) {
+		v4l2_err(&dev->v4l2_dev, "%s: Post wait: %d\n", __func__,
+			 status);
+		goto fail;
+	}
+
+	de->frame_buf = v4l2_m2m_cap_buf_detach(dev->m2m_dev, ctx->fh.m2m_ctx);
+	if (!de->frame_buf) {
+		v4l2_err(&dev->v4l2_dev, "%s: No detached buffer\n", __func__);
+		goto fail;
+	}
+
+	ctx->p2idx =
+		(ctx->p2idx + 1 >= RPIVID_P2BUF_COUNT) ? 0 : ctx->p2idx + 1;
+
+	// Enable the next setup if our Q isn't too big
+	if (atomic_add_return(1, &ctx->p2out) < RPIVID_P2BUF_COUNT) {
+		v4l2_m2m_buf_done_and_job_finish(dev->m2m_dev, ctx->fh.m2m_ctx,
+						 VB2_BUF_STATE_DONE);
+	}
+
+	rpivid_hw_irq_active2_claim(dev, &de->irq_ent, phase2_claimed, de);
+
+	return;
+
+fail:
+	dec_env_delete(ctx, de);
+	v4l2_m2m_buf_done_and_job_finish(dev->m2m_dev, ctx->fh.m2m_ctx,
+					 VB2_BUF_STATE_ERROR);
+}
+
+static void phase1_claimed(struct rpivid_dev *const dev, void *v)
+{
+	// (Re-)allocate PU/COEFF stream space
+	struct rpivid_dec_env *const de = v;
+	struct rpivid_ctx *const ctx = de->ctx;
+
+	const struct rpivid_gptr *const p2gptr = ctx->p2bufs + ctx->p2idx;
+	unsigned int pu_size;
+
+	de->pu_base_vc = p2gptr->addr;
+	de->pu_stride = rnd64(ctx->max_pu_msgs * 2 * de->PicWidthInCtbsY);
+	pu_size = de->pu_stride * de->PicHeightInCtbsY;
+
+	// Allocate all remaining space to coeff
+	de->coeff_base_vc = de->pu_base_vc + pu_size;
+	de->coeff_stride = ((p2gptr->size - pu_size) / de->PicHeightInCtbsY) &
+			   ~63; // Round down to multiple of 64
+
+	apb_write_vc_addr(dev, RPI_PUWBASE, de->pu_base_vc);
+	apb_write_vc_len(dev, RPI_PUWSTRIDE, de->pu_stride);
+	apb_write_vc_addr(dev, RPI_COEFFWBASE, de->coeff_base_vc);
+	apb_write_vc_len(dev, RPI_COEFFWSTRIDE, de->coeff_stride);
+
+	//	v4l2_info(&dev->v4l2_dev, "%s: Pre trigger\n", __func__);
+
+	// Trigger command FIFO
+	apb_write(dev, RPI_CFNUM, de->cmd_len);
+
+	// Claim irq
+	rpivid_hw_irq_active1_irq(dev, &de->irq_ent, cb_phase1, de);
+
+	// And start the h/w
+	apb_write_vc_addr(dev, RPI_CFBASE, de->cmd_copy_gptr->addr);
+	mb();
+}
+
+static void dec_state_delete(struct rpivid_ctx *const ctx)
+{
+	unsigned int i;
+	struct rpivid_dec_state *const s = ctx->state;
+
+	if (!s)
+		return;
+	ctx->state = NULL;
+
+	free_ps_info(s);
+
+	for (i = 0; i != HEVC_MAX_REFS; ++i)
+		aux_q_release(ctx, &s->ref_aux[i]);
+	aux_q_release(ctx, &s->frame_aux);
+
+	kfree(s);
+}
+
+static struct rpivid_dec_state *dec_state_new(void)
+{
+	struct rpivid_dec_state *s = kzalloc(sizeof(*s), GFP_KERNEL);
```

Comment:
i think it's preferred to use kzalloc directly

Reply:
I'm not sure if it's what you mean, but the function is actually giving us no gain.
```if (!s) return NULL``` is equivalent to ```return s```, and that is what happens in the success case. I'll drop the function.

Reply:
Yes, this is what i tried to say